addons/README.md

```diff
@@ -2,22 +2,40 @@
 
 ## Structure
 
-Each available add-on has a directory with subdirectories (<addon>/<version>) for each available version of the add-on.
-Each add-on is require to define at a minimum two files, `install.sh` and `Manifest`.
+Each available add-on has a directory with subdirectories (<addon>/<version>) for each available version of the add-on following the format `/addons/<name>/<version>`.
+Each add-on is requires to define at a minimum two files, `install.sh` and `Manifest`. You can check that each directory must have at least two files: `install.sh` and `Manifest`.
 
 The Manifest file specifies a list of assets required by the add-on.
 These will be downloaded during CI and saved to the add-on [directory](/ARCHITECTURE.md#directory-structure).
 
-The install.sh script can implement a set of [Lifecycle Hooks](#lifecycle-hooks).
+The install.sh script can implement a set of [Lifecycle Hooks](#lifecycle-hooks). For example, `/addons/weave/2.5.2/install.sh` defines the function named `weave`.
+The `install.sh` script must define a function named `<add-on>` that will perform the install.
 
 Any [other files](/ARCHITECTURE.md#directory-structure) in the <addon>/<version> subdirectory will be included in the package built for the add-on.
 The package will be built and uploaded to s3://[bucket]/[(dist\|staging)]/[kURL version]/<addon>-<version>.tar.gz during CI.
+It can be downloaded directly from S3 or by redirect from `https://kurl.sh/dist/<name>-<version>.tar.gz`.
+The built package will include the images from the Manifest saved as tar archives.
+
+Most add-ons include yaml files that will be applied to the cluster.
+These should be copied to the directory `kustomize/<name>` and applied with `kubectl apply -k` rather than applied directly with `kubectl apply -f`.
+This will allow users to easily review all applied yaml, add their own patches and re-apply after the script completes.
+
+The `install.sh` script may also define the functions `<name>_pre_init` and `<name>_join`.
+The pre_init function will be called prior to initializing the Kubernetes cluster.
+This may be used to modify the configuration that will be passed to `kubeadm init`.
+The join function will be called on remote nodes before they join the cluster and is useful for host configuration tasks such as loading a kernel module.
 
 ## Runtime
 
 During installation, upgrades and joining additional nodes, the installer will invoke a set of add-on lifecycle hooks.
```

Comment:
```suggestion
During installation, upgrades and joining additional nodes, the installer will first load all assets and container images from the add-on's directory and create the directory `$DIR/kustomize/<add-on>` for rendered add-on scripts.
It will then source the `<add-on>/install.sh` script and execute a set of add-on [lifecycle hooks](#lifecycle-hooks).

Functions relating to the add-on runtime can be found [here](https://github.com/replicatedhq/kurl/blob/master/scripts/common/addon.sh).
```